### PR TITLE
[Indexer] Recode state change set for state sync And provide state sync RPC

### LINKED ISSUE #1038: [Indexer] Record Changeset indexer table
Create a Changeset table for data recovery and state sync

### PR DESCRIPTION
**Pull request by baichuan3**
## Summary

Part of #1028 and follow up https://github.com/rooch-network/rooch/pull/1186

1. Recode state change set for state sync 
2. Provide state sync RPC, sync from asc by default
3. Add state sync test cases
4. Rename `key_hash ` to `key_str ` for leaf_states table

- Closes #1038

### CODE REVIEW

**Comment by baichuan3** [comment]
```
curl --location 'http://localhost:50051' \
--header 'Content-Type: application/json' \
--data '{
    "id": 101,
    "jsonrpc": "2.0",
    "method": "rooch_syncStates",
    "params": [
        null, "10", null
    ]
}'
```

```
{
    "jsonrpc": "2.0",
    "result": {
        "data": [
            {
                "tx_order": 0,
                "state_change_set": {
                    "new_tables": {},
                    "removed_tables": [],
                    "changes": {
                        "0x0000000000000000000000000000000000000000000000000000000000000000": {
                            "entries": {
                                "0x4b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c56": {
                                    "type": "new",
                                    "value": "0x4b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c564b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c5600447c72e15e318ef69ef0de731754a97120e4f1448a127bae8f82ef7000ffb1a3a6c051877fbbb3398ec27d9eca77658bcc93b63f5ecca2eaef5e5d0ffc1fb0ec",
                                    "value_type": "0x2::object::ObjectEntity<0x2::account_storage::AccountStorage>",
                                    "decoded_value": null
                                },
                                "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a": {
                                    "type": "modify",
                                    "value": "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a00000000000000000000000000000000000000000000000000000000000000000053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e0ce11a000000000000000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                },
                                "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd5764": {
                                    "type": "new",
                                    "value": "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd57644b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c560053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696ef41ef5622d5ec76b05000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                },
                                "0xf81628c3bf85c3fc628f29a3739365d4428101fbbecca0dcc7e3851f34faea6b": {
                                    "type": "modify",
                                    "value": "0xf81628c3bf85c3fc628f29a3739365d4428101fbbecca0dcc7e3851f34faea6b00000000000000000000000000000000000000000000000000000000000000030053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e0e526f6f63682047617320436f696e0352474312000010632d5ec76b050000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin::CoinInfo<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 2
                        },
                        "0x447c72e15e318ef69ef0de731754a97120e4f1448a127bae8f82ef7000ffb1a3": {
                            "entries": {
                                "0x52303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e743a3a4163636f756e74": {
                                    "type": "new",
                                    "value": "0x0100000000000000",
                                    "value_type": "0x3::account::Account",
                                    "decoded_value": null
                                },
                                "0x60303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e745f636f696e5f73746f72653a3a436f696e53746f726573": {
                                    "type": "new",
                                    "value": "0xad7d15099438f06bc6e5954089b798d6ebd4a8cf8d2201e29a7872e0a11fff93",
                                    "value_type": "0x3::account_coin_store::CoinStores",
                                    "decoded_value": null
                                },
                                "0x6c303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e745f61757468656e7469636174696f6e3a3a41757468656e7469636174696f6e4b657973": {
                                    "type": "new",
                                    "value": "0xa8b28ae0a36c51f6998eec632eb3b75777f49ae312cc0c17d75f20f61fb47135",
                                    "value_type": "0x3::account_authentication::AuthenticationKeys",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 3
                        },
                        "0xa6c051877fbbb3398ec27d9eca77658bcc93b63f5ecca2eaef5e5d0ffc1fb0ec": {
                            "entries": {
                                "0x0a6576656e745f74657374": {
                                    "type": "new",
                                    "value": "0x8502a11ceb0b0600000009010006020608030e10041e02052012073252088401400ac401070ccb011200000101010200030300010400000005000100010802030002090501010202040207080103000106080101050108000109000a6576656e745f7465737407636f6e74657874056576656e740d57697468647261774576656e7407436f6e746578740a656d69745f6576656e74046164647206616d6f756e740673656e64657204656d69744b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c560000000000000000000000000000000000000000000000000000000000000002000202060507030001040001070b002e11010b01120038000200",
                                    "value_type": "0x2::move_module::MoveModule",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 1
                        },
                        "0xa8b28ae0a36c51f6998eec632eb3b75777f49ae312cc0c17d75f20f61fb47135": {
                            "entries": {
                                "0xd001303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e745f61757468656e7469636174696f6e3a3a41757468656e7469636174696f6e4b65793c303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6e61746976655f76616c696461746f723a3a4e617469766556616c696461746f723e": {
                                    "type": "new",
                                    "value": "0x204b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c56",
                                    "value_type": "0x3::account_authentication::AuthenticationKey<0x3::native_validator::NativeValidator>",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 1
                        },
                        "0xad7d15099438f06bc6e5954089b798d6ebd4a8cf8d2201e29a7872e0a11fff93": {
                            "entries": {
                                "0x53303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e": {
                                    "type": "new",
                                    "value": "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd5764",
                                    "value_type": "0x2::object::ObjectID",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 1
                        }
                    }
                },
                "created_at": 0
            },
            {
                "tx_order": 1,
                "state_change_set": {
                    "new_tables": {},
                    "removed_tables": [],
                    "changes": {
                        "0x0000000000000000000000000000000000000000000000000000000000000000": {
                            "entries": {
                                "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a": {
                                    "type": "modify",
                                    "value": "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a00000000000000000000000000000000000000000000000000000000000000000053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e71171b000000000000000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                },
                                "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd5764": {
                                    "type": "modify",
                                    "value": "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd57644b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c560053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e8fe8f4622d5ec76b05000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        },
                        "0x447c72e15e318ef69ef0de731754a97120e4f1448a127bae8f82ef7000ffb1a3": {
                            "entries": {
                                "0x52303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e743a3a4163636f756e74": {
                                    "type": "modify",
                                    "value": "0x0200000000000000",
                                    "value_type": "0x3::account::Account",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        }
                    }
                },
                "created_at": 0
            },
            {
                "tx_order": 2,
                "state_change_set": {
                    "new_tables": {},
                    "removed_tables": [],
                    "changes": {
                        "0x0000000000000000000000000000000000000000000000000000000000000000": {
                            "entries": {
                                "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a": {
                                    "type": "modify",
                                    "value": "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a00000000000000000000000000000000000000000000000000000000000000000053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696ed64d1b000000000000000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                },
                                "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd5764": {
                                    "type": "modify",
                                    "value": "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd57644b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c560053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e2ab2f4622d5ec76b05000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        },
                        "0x447c72e15e318ef69ef0de731754a97120e4f1448a127bae8f82ef7000ffb1a3": {
                            "entries": {
                                "0x52303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e743a3a4163636f756e74": {
                                    "type": "modify",
                                    "value": "0x0300000000000000",
                                    "value_type": "0x3::account::Account",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        }
                    }
                },
                "created_at": 0
            },
            {
                "tx_order": 3,
                "state_change_set": {
                    "new_tables": {},
                    "removed_tables": [],
                    "changes": {
                        "0x0000000000000000000000000000000000000000000000000000000000000000": {
                            "entries": {
                                "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a": {
                                    "type": "modify",
                                    "value": "0x5abccdaadbbbc049420786674b8b899b0da4b498391626f2da59d97787c8c84a00000000000000000000000000000000000000000000000000000000000000000053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696e3b841b000000000000000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                },
                                "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd5764": {
                                    "type": "modify",
                                    "value": "0x9fdf6a44e40e34899dca64cddf78a4f8896e82d9b3b52d8c3757e193a2dd57644b1b5d2a62f3c9a19a5314167439945bf98fbc871e1402e36d5bad3382064c560053303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6761735f636f696e3a3a476173436f696ec57bf4622d5ec76b05000000000000000000000000000000000000000000000000",
                                    "value_type": "0x2::object::ObjectEntity<0x3::coin_store::CoinStore<0x3::gas_coin::GasCoin>>",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        },
                        "0x447c72e15e318ef69ef0de731754a97120e4f1448a127bae8f82ef7000ffb1a3": {
                            "entries": {
                                "0x52303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030333a3a6163636f756e743a3a4163636f756e74": {
                                    "type": "modify",
                                    "value": "0x0400000000000000",
                                    "value_type": "0x3::account::Account",
                                    "decoded_value": null
                                }
                            },
                            "size_increment": 0
                        }
                    }
                },
                "created_at": 0
            }
        ],
        "next_cursor": 3,
        "has_next_page": false
    },
    "id": 101
}
```

File: crates/rooch-indexer/migrations/2023-11-21-122443_leaf_states/up.sql [hunk 1]
```diff
@@ -2,12 +2,12 @@ CREATE TABLE leaf_states
 (
     id                 VARCHAR        NOT NULL      PRIMARY KEY,
     object_id          VARCHAR        NOT NULL,
-    key_hash           VARCHAR        NOT NULL,
+    key_str            VARCHAR        NOT NULL,
```
Comment (jolestar) [inline crates/rooch-indexer/migrations/2023-11-21-122443_leaf_states/up.sql:5]: `key` is a keyword for the SQL? How about `key_hex`?

File: crates/rooch-rpc-api/src/jsonrpc_types/state_view.rs [hunk 2]
```diff
@@ -102,9 +111,20 @@ impl From<Op<State>> for OpView<StateView> {
...
+            OpView::Modify(data) => Self::Modify(data.into()),
+            OpView::Delete => Self::Delete,
+        }
+    }
+}
+
 #[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
 pub struct TableChangeView {
     pub entries: BTreeMap<BytesView, OpView<StateView>>,
+    pub size_increment: i64,
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/state_view.rs:127]: Maybe this new field breaks the js sdk.

File: crates/rooch-indexer/src/indexer_reader.rs [hunk 3]
```diff
@@ -297,4 +299,64 @@ impl IndexerReader {
...
+            ",
+            where_clause, order_clause, limit,
+        );
+
+        tracing::debug!("sync states: {}", query);
+        let stored_state_change_sets = self
+            .inner_indexer_reader
+            .run_query(|conn| diesel::sql_query(query).load::<StoredStateChangeSet>(conn))?;
+
+        let result = stored_state_change_sets
```
Comment (jolestar) [inline crates/rooch-indexer/src/indexer_reader.rs:349]: Maybe we should merge all ChangeSets into one ChangeSet to reduce the size in the future.

File: crates/rooch-indexer/src/models/states.rs [hunk 4]
```diff
@@ -131,3 +135,43 @@ impl StoredLeafState {
...
+
+#[derive(Clone, Debug, Queryable, Insertable, QueryableByName)]
+#[diesel(table_name = state_change_sets)]
+pub struct StoredStateChangeSet {
+    /// The tx order of this transaction which produce the state change set
+    #[diesel(sql_type = diesel::sql_types::BigInt)]
+    pub tx_order: i64,
+    /// The state change set, json format
+    #[diesel(sql_type = diesel::sql_types::Text)]
+    pub state_change_set: String,
```
Comment (jolestar) [inline crates/rooch-indexer/src/models/states.rs:147]: If we store the whole `ChangeSet` into a json, how do we provide `ChangeSet` filter base on table?

**Comment by jolestar** [comment]
> Encouter an CI sdk test error, https://github.com/rooch-network/rooch/actions/runs/7044107582/job/19179560537?pr=1190,
> 
> ```
>  ❯ test/e2e/sdk.test.ts  (13 tests | 1 failed) 19010ms
>    ❯ test/e2e/sdk.test.ts > SDK > #sessionKey > Create session account with multi scopes should be ok
>      → status ABORTED of type Execution with sub status 327684
> ```
> 
> Please help to have a look. @jolestar @wow-sven

It is the SDK test case timestamp issue. The same error was encountered at #1187

File: crates/rooch-rpc-api/src/jsonrpc_types/state_view.rs (hunk 2 above)
Comment (jolestar) [reply crates/rooch-rpc-api/src/jsonrpc_types/state_view.rs:127]: Not this change breaks the SDK test. But the typescript `TableChangeView` also needs to be updated.

**Review by jolestar — APPROVED** [review]
I will merge this pr first.

File: crates/rooch-indexer/src/models/states.rs (hunk 4 above)
Comment (baichuan3) [reply crates/rooch-indexer/src/models/states.rs:147]: Split StateChangeSet into TableChangeSet dimensions to store?

Comment (jolestar) [reply crates/rooch-indexer/src/models/states.rs:147]: Yes, we can keep the table handle in this table. Then, the developer can filter ChangeSet by table handle. The 0x0 is the global Object table.

**Pull request merged**